SpriteFrameCache: take pixelFormat into account if specified

Shulepov opened the pull request:
TexturePacker has nice feature - specify pixelFormat for atlas (RGBA8888, RGBA4444, etc). It may help to save used storage space and game memory. But currently it's not used in cocos2d-x and textures always loaded in currently specified DefaultPixelFormat (usually RGBA8888).

Though, currently info about pixelFormat also not exported by TexturePacker, but it can be added.

Shulepov committed:
SpriteFrameCache: take pixelFormat into account if specified

CodeAndWeb commented:
Sounds like a good idea to me. We'll add the additional value with the next TexturePacker update.

ricardoquesada commented:
@Shulepov what is the best way to test this new API? could you attach a test case? thanks.

Shulepov commented:
@ricardoquesada ofc, will add some tests soon.
I think should be provided few sets of atlases with different specified pixelFormat and checked that target texture has corresponding pixel format.

ricardoquesada commented:
@Shulepov ok. thanks.

Shulepov committed:
Texture2D: RGB888 -> A8 conversion (white - full opacity, black - full transparency)

Shulepov committed:
Added aliases for PixelFormat names

Shulepov commented:
A8 textures are often represented as grayscale images where _black_ - full transparency, _white_ - full opacity. And them are firstly read by cocos2d-x as _Image_ with format RGB888. But there is no conversion RGB888 -> A8. So i have added one similar to RGB888 -> I8.

Sample image: https://i.imgsafe.org/4c94ea1.png

Shulepov committed:
Texture2D::getStringForFormat - added missed formats

Shulepov committed:
SpriteFrameCacheTest: test for PixelFormat

Shulepov commented:
Added test:
- added few atlases at _'ResourcesImages/sprite_frames_test'_ with different _pixelFormat_;
- during test load these atlases and check that corresponding texture has expected _PixelFormat_;

Wasn't sure what to show on the screen, so showing just some info about loaded textures - format name and size in KB.

Shulepov commented:
Whoops, forgot to add new test case for other platforms. Though i can add it to _jni_ and _CMakeLists_, but i don't have windows, so can't properly add it to _cpp-tests.vcxproj.filters_ for win32 and win10... (it requires _UniqueIdentifier_ for new group, not sure whether it must be generated only by IDE).

Shulepov committed:
Add SpriteFrameCacheTest to Android, Windows, Linux project files

minggo commented:
@ricardoquesada Could you please take a look of this issue? I will merge it if it is ready to be merged.

Thanks.

CodeAndWeb commented:
TexturePacker contains the format update. We've also added a flag for premultiplied alpha - maybe the code should check for this, too?

```
            <key>pixelFormat</key>
            <string>RGBA8888</string>
            <key>premultiplyAlpha</key>
            <false/>

```

What about PVRTC and others? Are they determined from the file directly?

ricardoquesada commented:
@minggo yep. it is ready to be merged.

@CodeAndWeb sorry, I didn't understand your question. What do you mean by "PVRTC and others" ? thanks.

CodeAndWeb commented:
The code covers RGBA8888,RGBA4444,RGB5A1,RGB565,A8,I8,AI88,BGRA8888,RGB888.

For other formats like PVRTC, ETC1,... it simply falls back to the previous way of handling textures - by just passing them to the TextureCache - not setting anything special.

I think this is ok - just wanted to make sure we don't overlook something.

minggo commented on tests/cpp-tests/Classes/SpriteFrameCacheTest/SpriteFrameCacheTest.cpp:
```diff
@@ -0,0 +1,85 @@
+    loadSpriteFrames("Images/sprite_frames_test/test_RGBA4444.plist", Texture2D::PixelFormat::RGBA4444);
+    loadSpriteFrames("Images/sprite_frames_test/test_RGBA5551.plist", Texture2D::PixelFormat::RGB5A1);
+#if CC_TARGET_PLATFORM == CC_PLATFORM_IOS || CC_TARGET_PLATFORM == CC_PLATFORM_MAC
+    loadSpriteFrames("Images/sprite_frames_test/test_PVRTC2.plist", Texture2D::PixelFormat::PVRTC2A);
```
The test case will crash on Mac.

minggo replied:
Will crash at `CC_ASSERT(texture->getPixelFormat() == expectedFormat);`

minggo commented on cocos/2d/CCSpriteFrameCache.cpp:
```diff
@@ -298,6 +298,58 @@ void SpriteFrameCache::addSpriteFramesWithDictionary(ValueMap& dictionary, Textu
+        {"AI88", Texture2D::PixelFormat::AI88},
+        {"ALPHA_INTENSITY", Texture2D::PixelFormat::AI88},
+        //{"BGRA8888", Texture2D::PixelFormat::BGRA8888}, no Image convertion RGBA -> BGRA
+        {"RGB888", Texture2D::PixelFormat::RGB888}
```
Why only these pixel format should change Texture2D's default pixel format before creating textures?

minggo commented on tests/cpp-tests/Classes/SpriteFrameCacheTest/SpriteFrameCacheTest.cpp:
```diff
@@ -0,0 +1,85 @@
+    Texture2D *texture = spriteFrame->getTexture();
+    const ssize_t bitsPerKB = 8 * 1024;
+    const double memorySize = 1.0 * texture->getBitsPerPixelForFormat() * texture->getContentSizeInPixels().width * texture->getContentSizeInPixels().height / bitsPerKB;
+    CC_ASSERT(texture->getPixelFormat() == expectedFormat);
```
`texture->getPixelFormat() != expectedFormat` when `expectedFormat` is `Texture2D::PixelFormat::PVRTCXX`

minggo replied:
@Shulepov I will comment this line if you don't have different opinion.

minggo replied:
i comment this line in #15356

Shulepov replied:
@minggo, sorry for the late response. Take a look at PR #15700

minggo replied:
Thanks. I merged it.

minggo commented:
@Shulepov Please take a look.
